<p>For the demo, we provide two datasets namely 1) case-study and 2) 1000 Genomes datasets that allows to perform pedigree analysis, case-control studies or comparisons across populations.<br></p>

<p><b>Case-study dataset:</b> The data from the study "A novel KRT71 variant in curly‚Äêcoated dogs" is used for the case-study. This includes 1.1 million variants from 330 samples with one affected sample and 329 unaffected samples.<br> </p>

<p><b>1000 Genomes dataset: </b> The 1000 Genomes Phase 3 dataset includes 1.2 million variants from 2,504 individuals across 26 populations.</p>


<p><b>Workflow for Default datasets:</b><br></p>

<b>Step 1)</b> Choose Variant Data (VCF) and select Case study/1000 Genomes variant database. <br></p>

<p><b>Step 2)</b> Choose Phenotype File and select Case study/1000 Genomes phenotype file. <br></p>

<p><b>Step 3)</b> Choose filtering module and click "Filter". The status message is displayed at the bottom of the page while the querying is performed and displays the results after completion.<br></p>


<p><b>Workflow for user uplaoded dataset:</b><br></p>

<b>Step 1)</b> Choose Variant Data (VCF) and upload GQT indexed VCF files. <br></p>

<p><b>Step 2)</b> Choose Phenotype File and "Upload PED" <b>-></b> "Check Samples" <b>-></b> "CreateDB"<br></p>

<p>In step 2, the user uploads the PED file, checks the phenotype status of samples from the data table and proceed to create sample database."CreateDB" will show the status message at the bottom of the page when the database is being created. This step has to be done once whenever the user uploads a new phenotype file. After creating the GQT sample database variant filtering can be done consecutively in one or more modules.</p>

<p><b>Step 3)</b> Choose filtering module and click "Filter". The status message is displayed at the bottom of the page while the querying is performed and displays the results after completion.<br></p>



