<b>webGQT</b> is a graphical web user interface to perform genotype querying large-scale genotype data from thousands to millions of genomes.
webGQT requires a PED file with sample information and the corresponding GQT indexed VCF/BCF files. This application includes four modules: pedigree analysis, case-control studies, population frequency studies and variant count module with pre-built queries to filter candidate variants for genetic disease studies from millions of variants.
webGQT enable researchers with less IT or applied bioinformatics experience to explore massive datasets with few mouse clicks gaining substantial speed over the exising methods as demonstrated in Layer et.al.</p>

Pedigree analysis module allows to filter variants by mendelian inheritance disease models namely recessive, dominant and de novo models among big cohort of samples with the affection status in the phenotype file. Case-control studies module allows to filter case specific/shared variants with the case and control status defined in the phenotype file. Population module filters the variants between samples of two populations and specific genders defined in the phenotype file. Variant count module gives the count of total number of variants in specific sample.</p>

<b>Results: </b>webGQT outputs the filtered variants in each module in tabular and standard VCF format. The VCF format output can be used for further downstream analysis with external programs eg: bcftools. Inaddition, variant summary table and a histogram is generated using the gene annotations in the input VCF file.</p>
