<p>For the demo, we provide a small number of variants from 1000 Genomes datasets that allows to perform all the filtering tasks across pedigree, case-control and population modules.<br></p>


<p><b>1000 Genomes dataset: </b> The demo dataset from 1000 Genomes Phase 3 dataset includes 768 variants from 2,504 individuals across 26 populations.</p>

Implementing webGQT is a simple three step process and the following documentation describes the step-by-step work flow for the defualt and the user uploaded datasets.
<p><b>Workflow for Default datasets:</b><br></p>

<b>Step 1)</b> Choose Variant Data (VCF) and select 1000 Genomes variant database. <br></p>

<p><b>Step 2)</b> Choose Phenotype File and select 1000 Genomes phenotype file. <br></p>

<p><b>Step 3)</b> Choose filtering module and click "Filter". The status message is displayed at the bottom of the page while the querying is performed and displays the results after completion.<br></p>


<p><b>Workflow for user uplaoded dataset:</b><br></p>

<b>Step 1)</b> Choose Variant Data (VCF) and upload GQT indexed VCF files. <br></p>

<p><b>Step 2)</b> Choose Phenotype File and "Upload PED" <b>-></b> "Check Samples" <b>-></b> "CreateDB"<br></p>

<p>In step 2, the user uploads the PED file, checks the phenotype status of samples from the data table and subsequently creates sample database."CreateDB" will show the status message at the bottom of the page when the database is being created. This step has to be done once whenever the user uploads a new phenotype file. After creating the GQT sample database variant filtering can be done consecutively in one or more modules.</p>

<p><b>Step 3)</b> Choose filtering module and click "Filter". The status message is displayed at the bottom of the page while the querying is performed and displays the results after completion.<br></p>



